torch/_functorch/compile_utils.py

```diff
@@ -50,6 +50,14 @@ def fx_graph_cse(fx_g: torch.fx.graph.Graph):
     )
 
     compute_mutation_region_ids(fx_g)  # type: ignore[arg-type]
+
+    # Make a list of nodes that are directly returned from the output, which we'll skip
+    # when pruning.  This prevents us from deduplicating returned tensors which have
+    # experienced identical operations, but are separate data structures in eager mode.
+    output_node: fx.Node = list(fx_g.nodes)[-1]
+    assert output_node.op == "output"
+    inputs_to_output_node = output_node.all_input_nodes
```

Comment:
If we want to handle the aliasing case that @zou3519 mentioned above - can't we just apply the same treatment to any nodes that alias outputs as well?
```
out_storages = set([StorageWeakRef(n.meta['val'].untyped_storage()) for n in output_node.args if 'val' in n.meta and isinstance(n.meta['va'], torch.Tensor)])
nodes_that_alias_outputs = set()
for n in fx_g.graph.nodes:
    if 'val' in n.meta and isinstance(n.meta['val'], torch.Tensor):
        if StorageWeakRef(n.meta['val'].untyped_storage()) in out_storages:
            nodes_that_alias_outputs.add(n)

# nodes in "nodes_that_alias_outputs" should not be CSE'd
```

Reply:
@bdhirsh I can try that! I was wondering if the long-term answer might look something like adding this check for the return of `untyped_storage` into the parameter set we check to de-duplicate nodes. That might be too pessimistic in terms of what we can eliminate, though.

Reply:
@bdhirsh Good call! @zou3519 this suggestion appears to resolve the issue you identified; I've pushed it and a corresponding update to the repro tests.